kyo-prelude/shared/src/test/scala/kyo2/EnvTest.scala

```diff
@@ -0,0 +1,314 @@
+                        kyo2.Env.get[Service2].map(_(i))
+                    }
+                val v: Int < (Env[Service1] & Env[Service2] & Abort[None.type]) = a
+                "same handling order" in {
+                    val b = kyo2.Env.run(service2)(v)
+                    val c = kyo2.Env.run(service1)(b)
+                    assert(
+                        kyo2.Abort.run(c).eval == Right(3)
+                    )
+                }
+                "reverse handling order" in {
+                    val b = kyo2.Env.run(service1)(v)
+                    val c = kyo2.Env.run(service2)(b)
+                    assert(
+                        kyo2.Abort.run(c).eval == Right(3)
+                    )
+                }
+                "dependent services" in {
+                    val v2: Int < (Env[Service2] & Abort[None.type]) = kyo2.Env.run(service1)(v)
+                    assert(
+                        kyo2.Abort.run(kyo2.Env.run(service2)(v2)).eval ==
+                            Right(3)
+                    )
+                }
+            }
+        }
+    }
+
+    "runTypeMap" - {
+        "providing env maps" in {
+            val kyo =
+                for
+                    string <- kyo2.Env.get[String]
+                    int    <- kyo2.Env.get[Int]
+                    bool   <- kyo2.Env.get[Boolean]
+                yield (string, int, bool)
+
+            val envMap = TypeMap("Hello", 123, true)
+            assert(
+                kyo2.Env.runTypeMap(envMap)(kyo).eval == ("Hello", 123, true)
+            )
+        }
+
+        "leaving off one service" in {
+            val kyo =
+                for
+                    string <- kyo2.Env.get[String]
+                    int    <- kyo2.Env.get[Int]
+                    bool   <- kyo2.Env.get[Boolean]
+                yield (string, int, bool)
+
+            val envMap: TypeMap[String & Int]                      = TypeMap("Hello", 123)
+            val withTypeMap: (String, Int, Boolean) < Env[Boolean] = kyo2.Env.runTypeMap(envMap)(kyo)
+            val withBool: (String, Int, Boolean) < Any             = kyo2.Env.run(true)(withTypeMap)
+            assert(
+                withBool.eval == ("Hello", 123, true)
+            )
+        }
+
+        "multiple provide calls" in {
+            val kyo =
+                for
+                    string <- kyo2.Env.get[String]
+                    int    <- kyo2.Env.get[Int]
+                    bool   <- kyo2.Env.get[Boolean]
+                yield (string, int, bool)
+
+            val stringTypeMap = TypeMap("Hello")
+            val intTypeMap    = TypeMap(123)
+            val boolTypeMap   = TypeMap(true)
+            assert(
+                kyo2.Env.run(true)(
+                    kyo2.Env.runTypeMap(stringTypeMap)(kyo2.Env.runTypeMap(intTypeMap)(kyo2.Env.runTypeMap(boolTypeMap)(kyo)))
+                ).eval == ("Hello", 123, true)
+            )
+        }
+
+        "providing the wrong env map" in {
```

Comment:
Note how the scenarios where the effect getting handled isn't pending are now supported and correctly leave effects still pending. This is possible because the computation gets nested via `AnyVal` boxing when the effects have a mismatch.